### PR TITLE
jewel: rbd: rbd_clone_copy_on_read ineffective with exclusive-lock

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/19174

### CODE REVIEW

**Comment by smithfarm** [comment]
```
In file included from ../src/gmock/gtest/include/gtest/gtest.h:1929:0,
                 from ./test/librbd/test_fixture.h:7,
                 from ./test/librbd/test_mock_fixture.h:7,
                 from test/librbd/image/test_mock_RefreshRequest.cc:4:
test/librbd/image/test_mock_RefreshRequest.cc: In member function ‘virtual void librbd::image::TestMockImageRefreshRequest_ExclusiveLockWithCoR_Test::TestBody()’:
test/librbd/image/test_mock_RefreshRequest.cc:710:36: error: ‘class librbd::Operations<librbd::ImageCtx>’ has no member named ‘update_features’
     ASSERT_EQ(0, ictx->operations->update_features(RBD_FEATURE_JOURNALING,
                                    ^
../src/gmock/gtest/include/gtest/gtest_pred_impl.h:77:52: note: in definition of macro ‘GTEST_ASSERT_’
   if (const ::testing::AssertionResult gtest_ar = (expression)) \
                                                    ^
../src/gmock/gtest/include/gtest/gtest_pred_impl.h:166:3: note: in expansion of macro ‘GTEST_PRED_FORMAT2_’
   GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_FATAL_FAILURE_)
   ^
../src/gmock/gtest/include/gtest/gtest.h:1993:3: note: in expansion of macro ‘ASSERT_PRED_FORMAT2’
   ASSERT_PRED_FORMAT2(::testing::internal:: \
   ^
../src/gmock/gtest/include/gtest/gtest.h:2011:32: note: in expansion of macro ‘GTEST_ASSERT_EQ’
 # define ASSERT_EQ(val1, val2) GTEST_ASSERT_EQ(val1, val2)
                                ^
test/librbd/image/test_mock_RefreshRequest.cc:710:5: note: in expansion of macro ‘ASSERT_EQ’
     ASSERT_EQ(0, ictx->operations->update_features(RBD_FEATURE_JOURNALING,
     ^
test/librbd/image/test_mock_RefreshRequest.cc:715:36: error: ‘class librbd::Operations<librbd::ImageCtx>’ has no member named ‘update_features’
     ASSERT_EQ(0, ictx->operations->update_features(RBD_FEATURE_FAST_DIFF,
                                    ^
../src/gmock/gtest/include/gtest/gtest_pred_impl.h:77:52: note: in definition of macro ‘GTEST_ASSERT_’
   if (const ::testing::AssertionResult gtest_ar = (expression)) \
                                                    ^
../src/gmock/gtest/include/gtest/gtest_pred_impl.h:166:3: note: in expansion of macro ‘GTEST_PRED_FORMAT2_’
   GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_FATAL_FAILURE_)
   ^
../src/gmock/gtest/include/gtest/gtest.h:1993:3: note: in expansion of macro ‘ASSERT_PRED_FORMAT2’
   ASSERT_PRED_FORMAT2(::testing::internal:: \
   ^
../src/gmock/gtest/include/gtest/gtest.h:2011:32: note: in expansion of macro ‘GTEST_ASSERT_EQ’
 # define ASSERT_EQ(val1, val2) GTEST_ASSERT_EQ(val1, val2)
                                ^
test/librbd/image/test_mock_RefreshRequest.cc:715:5: note: in expansion of macro ‘ASSERT_EQ’
     ASSERT_EQ(0, ictx->operations->update_features(RBD_FEATURE_FAST_DIFF,
     ^
test/librbd/image/test_mock_RefreshRequest.cc:720:36: error: ‘class librbd::Operations<librbd::ImageCtx>’ has no member named ‘update_features’
     ASSERT_EQ(0, ictx->operations->update_features(RBD_FEATURE_OBJECT_MAP,
                                    ^
../src/gmock/gtest/include/gtest/gtest_pred_impl.h:77:52: note: in definition of macro ‘GTEST_ASSERT_’
   if (const ::testing::AssertionResult gtest_ar = (expression)) \
                                                    ^
../src/gmock/gtest/include/gtest/gtest_pred_impl.h:166:3: note: in expansion of macro ‘GTEST_PRED_FORMAT2_’
   GTEST_PRED_FORMAT2_(pred_format, v1, v2, GTEST_FATAL_FAILURE_)
   ^
../src/gmock/gtest/include/gtest/gtest.h:1993:3: note: in expansion of macro ‘ASSERT_PRED_FORMAT2’
   ASSERT_PRED_FORMAT2(::testing::internal:: \
   ^
../src/gmock/gtest/include/gtest/gtest.h:2011:32: note: in expansion of macro ‘GTEST_ASSERT_EQ’
 # define ASSERT_EQ(val1, val2) GTEST_ASSERT_EQ(val1, val2)
                                ^
test/librbd/image/test_mock_RefreshRequest.cc:720:5: note: in expansion of macro ‘ASSERT_EQ’
     ASSERT_EQ(0, ictx->operations->update_features(RBD_FEATURE_OBJECT_MAP,
     ^
test/librbd/image/test_mock_RefreshRequest.cc:730:37: error: ‘expect_get_group’ was not declared in this scope
   expect_get_group(mock_image_ctx, 0);
                                     ^
make[4]: *** [test/librbd/image/unittest_librbd-test_mock_RefreshRequest.o] Error 1
```

**Comment by smithfarm** [comment]
@dillaman Could you suggest a fix for this build failure, caused by the absence of 93cf63f993a in jewel? (If the PR is not salvageable in its current form, feel free to do whatever you like with it.)

File: src/test/librbd/image/test_mock_RefreshRequest.cc [hunk 1]
```diff
@@ -682,6 +687,58 @@ TEST_F(TestMockImageRefreshRequest, JournalDisabledByPolicy) {
   ASSERT_EQ(0, ctx.wait());
 }
 
+TEST_F(TestMockImageRefreshRequest, ExclusiveLockWithCoR) {
+  REQUIRE_FEATURE(RBD_FEATURE_EXCLUSIVE_LOCK);
+
```
Comment (dillaman) [inline src/test/librbd/image/test_mock_RefreshRequest.cc]: Add ```REQUIRE(!is_feature_enabled(RBD_FEATURE_OBJECT_MAP | RBD_FEATURE_FAST_DIFF | RBD_FEATURE_JOURNALING))``` here

File: src/test/librbd/image/test_mock_RefreshRequest.cc [hunk 2]
```diff
@@ -682,6 +687,58 @@ TEST_F(TestMockImageRefreshRequest, JournalDisabledByPolicy) {
...
+  librbd::ImageCtx *ictx;
+  ASSERT_EQ(0, open_image(m_image_name, &ictx));
+
+  MockRefreshImageCtx mock_image_ctx(*ictx);
+  MockRefreshParentRequest mock_refresh_parent_request;
+
+  MockExclusiveLock mock_exclusive_lock;
+  mock_image_ctx.exclusive_lock = &mock_exclusive_lock;
+
+  if (ictx->test_features(RBD_FEATURE_JOURNALING)) {
```
Comment (dillaman) [inline src/test/librbd/image/test_mock_RefreshRequest.cc]: Yank these three feature conditions now that the test has a guard

File: src/test/librbd/image/test_mock_RefreshRequest.cc [hunk 3]
```diff
@@ -682,6 +687,58 @@ TEST_F(TestMockImageRefreshRequest, JournalDisabledByPolicy) {
...
+                                                   false));
+  }
+
+  expect_op_work_queue(mock_image_ctx);
+  expect_test_features(mock_image_ctx);
+
+  InSequence seq;
+  expect_get_mutable_metadata(mock_image_ctx, 0);
+  expect_get_flags(mock_image_ctx, 0);
+  expect_get_group(mock_image_ctx, 0);
```
Comment (dillaman) [inline src/test/librbd/image/test_mock_RefreshRequest.cc]: Remove this line

**Comment by smithfarm** [comment]
Jenkins re-test this please

**Comment by smithfarm** [comment]
@dillaman Thanks, that worked.

**Comment by smithfarm** [comment]
test this please

**Comment by smithfarm** [comment]
@dillaman This passed an rbd suite at http://tracker.ceph.com/issues/20613#note-31 (there was 1 failure that passed on rerun).

The rados run is putting up a bit of a fight; there are a couple failures but they don't appear to be reliably reproducible or related to rbd (or this PR, but what do I know?): http://tracker.ceph.com/issues/20613#note-25

I also ran three different upgrade suites; again, some failures failures that were either not reproducible or could be "hand-waved away".

OK to merge?

**Pull request merged**